### PR TITLE
qa: Enable RPC test execution to be overridden from Python

### PR DESCRIPTION
**Pull request by str4d**
`run_tests` now takes a (subclass of) `RPCTestHandler` as its first argument, and returns `True` if all tests passed instead of calling `sys.exit`. This enables RPC tests to be run from Python and the execution of individual tests to be customised:

```python
import importlib
import sys
sys.path.append('qa/pull-tester')
rpc_tests = importlib.import_module('rpc-tests')

src_dir = '.'
build_dir = '.'
exeext = ''

class MyTestHandler(rpc_tests.RPCTestHandler):
    def start_test(self, args, stdout, stderr):
        print('Starting test!')
        return subprocess.Popen(
            args,
            universal_newlines=True,
            stdout=stdout,
            stderr=stderr)

test_list = ['test_to_run.py']

all_passed = rpc_tests.run_tests(MyTestHandler, test_list, src_dir, build_dir, exeext)
```

### CODE REVIEW

**Comment by yasserisa** [comment]
I think the solution is great

From what I understand (please correct me if I'm wrong) with this example python script we can have there the list of tests to run and run the custom script to handle it with Kubernetes.

e.g:
```python
import importlib
import sys
sys.path.append('qa/pull-tester')
rpc_tests = importlib.import_module('rpc-tests')

src_dir = '.'
build_dir = '.'
exeext = ''

class MyTestHandler(rpc_tests.RPCTestHandler):
    def start_test(self, args, stdout, stderr):
        print('Starting test!')
        return subprocess.Popen(
            "kuberntes_script.sh" + args,
            universal_newlines=True,
            stdout=stdout,
            stderr=stderr)

test_list = ['test_to_run.py', 'test_to_run_2.py']

all_passed = rpc_tests.run_tests(MyTestHandler, test_list, src_dir, build_dir, exeext)
```

If this is correct. I have the following idea:

The list of tests is hardcoded in `rpc-tests.py`.

But to keep the list consistent, we should get that list of tests and add them to the `test_list` variable in the python example.

It occurs to me that maybe we could have the test list in a **txt** file. This way, when we want to add/remove tests, we simply modify the **txt file**. This way we maintain consistency to be called from this subclass or from `rpc-tests.py` itself.

What do you think about having this list separately?

**Comment by str4d** [comment]
It's even simpler: the tests are in constants in `rpc-tests.py`, which you can access directly.

I spent a bit of time migrating most of the Bash script from #6247 to work with the changes in this PR, as an example of what I mean:
```python
#!/usr/bin/env python3
import importlib
import subprocess
import sys
sys.path.append('qa/pull-tester')
rpc_tests = importlib.import_module('rpc-tests')

from random import randbytes

kube_job_template = '''
apiVersion: batch/v1
kind: Job
metadata:
  name: {random}
  namespace: pipelines
spec:
  ttlSecondsAfterFinished: 600
  backoffLimit: 10
  template:
    metadata:
      name: {random}
    spec:
      initContainers:
      - name: copy-artifact
        image: google/cloud-sdk
        volumeMounts:
        - name: workspace
          mountPath: /workspace
        command:
          - bash
        args:
          - -c
          - |
            gsutil -m cp gs://{artifacts_url}-tekton-artifacts/{pipelineRun_uid}-rpc.tar /workspace/
            cd /workspace
            tar -xvf {pipelineRun_uid}-rpc.tar
      containers:
      - name: rpc
        image: {build_container}
        resources:
          limits:
            memory: "4G"
            cpu: "4"
          requests:
              memory: "1G"
              cpu: "0.7"
        env:
        - name: ZCASHD
          value: /workspace/{pipelineRun_uid}/source/src/zcashd
        command:
          - bash
          - "-c"
          - |
            cp -a /home/.zcash-params /root/
            {test_cmd}
        volumeMounts:
        - name: workspace
          mountPath: /workspace
      serviceAccount: sa-github-gcr-gcs
      serviceAccountName: sa-github-gcr-gcs
      volumes:
        - name: workspace
          emptyDir: {{}}
      restartPolicy: OnFailure
      nodeSelector:
        node_pool: intensive-node-pool-c2
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
              - matchExpressions:
                  - key: node_pool
                    operator: In
                    values:
                      - intensive-node-pool-c2
      tolerations:
        - key: intensive-node-pool-c2
          operator: Exists
          effect: NoSchedule
'''

class KubeTestHandler(rpc_tests.RPCTestHandler):
    def start_test(self, args, stdout, stderr):
        # Maybe use the kubernetes Python library here, or:
        random = randbytes(6).hex()
        kube_job = random + '.yml'
        with open(kube_job, 'w') as f:
            f.write(kube_job_template.format(
                random = random,
                artifacts_url = 'ARTIFACTS_URL', # TODO: Maybe filled in by Tekton config?
                pipelineRun_uid = 'PIPELINERUN_UID',
                build_container = 'BUILD_CONTAINER',
                test_cmd = ' '.join(args),
            ))

        return subprocess.Popen(
            ['run_kube_job.sh', kube_job],
            universal_newlines=True,
            stdout=stdout,
            stderr=stderr)

def main():
    src_dir = '.'
    build_dir = '.'
    exeext = ''
    jobs = 150

    test_list = rpc_tests.SERIAL_SCRIPTS + rpc_tests.BASE_SCRIPTS + rpc_tests.ZMQ_SCRIPTS

    all_passed = rpc_tests.run_tests(
        KubeTestHandler,
        test_list,
        src_dir,
        build_dir,
        exeext,
        jobs,
    )

    # TODO: Do something with the result
    print(all_passed)

if __name__ == '__main__':
    main()
```

**Review by nuttycom — APPROVED** [review]
utACK

**Review by yasserisa — APPROVED** [review]
Tested ACK

**Comment by mdr0id** [comment]
Tested ACK in cluster

**Review by mdr0id — APPROVED** [review]

File: qa/pull-tester/rpc-tests.py [hunk 1]
```diff
@@ -376,10 +392,9 @@ def get_next(self):
             log_stderr = tempfile.SpooledTemporaryFile(max_size=2**16)
             self.jobs.append((t,
                               time.time(),
-                              subprocess.Popen((self.tests_dir + t).split() + self.flags + port_seed,
-                                               universal_newlines=True,
-                                               stdout=log_stdout,
-                                               stderr=log_stderr),
+                              self.start_test((self.tests_dir + t).split() + self.flags + port_seed,
```
Comment (daira) [inline qa/pull-tester/rpc-tests.py:395]: Pre-existing issue: consider using `os.path.join` instead of concatenating paths as strings.

**Review by daira — COMMENTED** [review]
Post-hoc ACK, with minor comment about a pre-existing issue.

**Pull request merged**